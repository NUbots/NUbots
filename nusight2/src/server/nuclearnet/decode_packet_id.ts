import { NUClearNetPacket } from "nuclearnet.js";
import { Reader } from "protobufjs/minimal";

import { messageFieldsIndex } from "../../shared/messages/fields_index";

/**
 * Decode and retrieve the id field from the given packet.
 * Returns a default id of 0 if the packet does not have an id field.
 */
export function decodePacketId(typeName: string, packet: NUClearNetPacket) {
  const defaultId = 0;

  const idField = messageFieldsIndex[typeName]?.id;
  if (idField === undefined) {
    return defaultId;
  }

  const reader = Reader.create(packet.payload);
  const end = reader.len;

  while (reader.pos < end) {
    // Tag is of the format: (fieldNumber << 3) | wireType
    // See https://developers.google.com/protocol-buffers/docs/encoding#structure
    const tag = reader.uint32();

    const fieldNumber = tag >>> 3;
    const wireType = tag & 7;

    if (fieldNumber === idField.id && wireType === idField.wireType) {
      // Assumes id fields are always 32 or 64-bit unsigned ints
      const id = idField.type === "uint32" ? reader.uint32() : reader.uint64();
      return id;
    }

    reader.skipType(wireType);
  }

  return defaultId;
}
