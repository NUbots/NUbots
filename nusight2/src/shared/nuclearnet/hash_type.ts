import XXH from "xxhashjs";

// Key for hashing NUClear types and subtypes.
const hashKey = 0x4e55436c;

/** Hash the given message type using the same algorithm used in NUClear */
export function hashType(type: string): Buffer {
  // Matches hashing implementation from NUClearNet
  // See https://github.com/Fastcode/NUClearNet.js/blob/857ed674/src/NetworkBinding.cpp#L32
  let hashString: string = XXH.h64(type, hashKey).toString(16);
  // The hash string may truncate if it's smaller than 16 characters so we pad it with 0s
  hashString = ("0".repeat(16) + hashString).slice(-16);

  return Buffer.from((hashString.match(/../g) as string[]).reverse().join(""), "hex");
}

/** Hash the given message subtype using the same algorithm used in NUClear */
export function hashSubtype(type: string): number {
  return XXH.h32(type, hashKey).toNumber();
}
