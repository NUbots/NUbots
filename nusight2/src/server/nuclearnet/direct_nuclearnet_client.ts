import { NUClearNet } from "nuclearnet.js";
import { NUClearNetOptions } from "nuclearnet.js";
import { NUClearNetPacket } from "nuclearnet.js";
import { NUClearNetSend } from "nuclearnet.js";

import { NUClearPacketListener } from "../../shared/nuclearnet/nuclearnet_client";
import { NUClearEventListener } from "../../shared/nuclearnet/nuclearnet_client";
import { NUClearNetClient } from "../../shared/nuclearnet/nuclearnet_client";

import { decodePacketId } from "./decode_packet_id";
import { parseEventString } from "./parse_event_string";

/**
 * A thin adapter around the real NUClearNet which implements the NUClearNetClient interface.
 */
export class DirectNUClearNetClient implements NUClearNetClient {
  constructor(private nuclearNetwork: NUClearNet) {}

  static of(): DirectNUClearNetClient {
    const nuclearNetwork = new NUClearNet();
    return new DirectNUClearNetClient(nuclearNetwork);
  }

  connect(options: NUClearNetOptions): () => void {
    this.nuclearNetwork.connect(options);
    return () => this.nuclearNetwork.disconnect();
  }

  onJoin(cb: NUClearEventListener): () => void {
    this.nuclearNetwork.on("nuclear_join", (peer) => {
      cb({ ...peer, type: "nuclearnet-peer" });
    });
    return () => this.nuclearNetwork.removeListener("nuclear_join", cb);
  }

  onLeave(cb: NUClearEventListener): () => void {
    this.nuclearNetwork.on("nuclear_leave", (peer) => {
      cb({ ...peer, type: "nuclearnet-peer" });
    });
    return () => this.nuclearNetwork.removeListener("nuclear_leave", cb);
  }

  on(event: string, cb: NUClearPacketListener): () => void {
    const { type, subtype } = parseEventString(event);

    // Filter out packets that don't match the given subtype if one was specified
    const handler =
      subtype !== undefined
        ? (packet: NUClearNetPacket) => {
            const id = decodePacketId(type, packet);
            if (id === subtype) {
              cb(packet, { subtype });
            }
          }
        : cb;

    this.nuclearNetwork.on(type, handler);
    return () => this.nuclearNetwork.removeListener(type, handler);
  }

  onPacket(cb: NUClearPacketListener): () => void {
    return this.on("nuclear_packet", cb);
  }

  send(options: NUClearNetSend): void {
    this.nuclearNetwork.send(options);
  }
}
