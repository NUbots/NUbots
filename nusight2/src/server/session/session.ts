import { NUClearNetSend } from "nuclearnet.js";

import { compose } from "../../shared/base/compose";
import { NUClearNetClient } from "../../shared/nuclearnet/nuclearnet_client";
import { createFilePicker } from "../file_picker/create";
import { createNbsScrubber } from "../nbs_scrubber/create";
import { ScrubberSet } from "../nbs_scrubber/scrubber_set";
import { ClientConnection } from "../web_socket/client_connection";

import { NUsightSessionClient } from "./session_client";
import { NUsightSessionNetwork } from "./session_network";

type ClientPacketCallback = (client: NUsightSessionClient, packet: NUClearNetSend) => void;
type ServerModuleCreator = (session: NUsightSession) => () => void;

/** Used for generating IDs for session clients */
let nextClientId = 1;

/**
 * Represents a "session": a collection of related NUsight browser clients
 * that share a common set of NBS scrubbers.
 */
export class NUsightSession {
  /** The browser clients currently connected to this session */
  readonly clients = new Set<NUsightSessionClient>();

  /**
   * A NUClearNet client connection used for sending and receiving messages
   * by clients in this session.
   */
  readonly nuclearnetClient: NUClearNetClient;

  /** Holds the scrubbers loaded by clients in this session */
  readonly scrubberSet: ScrubberSet;

  /**
   * The network for this session: allows for exchanging messages
   * between the server, clients in the session, and NUClearNet
   */
  readonly network: NUsightSessionNetwork;

  /** Map of client packet event names to listener callbacks */
  private clientPacketOns = new Map<string, Set<ClientPacketCallback>>();

  /** Clean up functions for the connected clients, run when a client disconnects */
  private clientCleanups = new Map<NUsightSessionClient, () => void>();

  /** Clean up function for modules installed into the session */
  private cleanUpModules: () => void;

  /** Create a new NUsightSession with the given NUClearNet connection */
  constructor(nuclearnetClient: NUClearNetClient, moduleCreators: ServerModuleCreator[]) {
    this.nuclearnetClient = nuclearnetClient;
    this.scrubberSet = ScrubberSet.of();
    this.network = new NUsightSessionNetwork(this);

    // Install the server-side modules for the session
    this.cleanUpModules = compose(moduleCreators.map((createModule) => createModule(this)));
  }

  static of(nuclearnetClient: NUClearNetClient) {
    return new NUsightSession(nuclearnetClient, [createNbsScrubber, createFilePicker]);
  }

  /** Add a client for the given connection to the session */
  addClient(clientConnection: ClientConnection): Readonly<NUsightSessionClient> {
    // Create and add a client for the connection
    const client = NUsightSessionClient.of(nextClientId++, clientConnection, {
      nuclearnetClient: this.nuclearnetClient,
      scrubberSet: this.scrubberSet,
      session: this,
    });

    this.clients.add(client);

    // Start processing messages for the client
    const stopProcessing = client.startProcessing();

    // Create a cleanup function for the client
    const cleanupClient = () => {
      stopProcessing();
      this.clients.delete(client);
      this.clientCleanups.delete(client);
    };

    // Clean up when the client disconnects
    clientConnection.onDisconnect(cleanupClient);

    // Store the cleanup function so we can call it when the client is removed
    // or when the session is destroyed
    this.clientCleanups.set(client, cleanupClient);

    // Send the client all currently loaded scrubbers in this session
    for (const scrubber of this.scrubberSet.scrubbers.values()) {
      clientConnection.send("nuclear_join", scrubber.peer);
    }

    return client;
  }

  /** Close the session and clean up all client connections and scrubbers */
  destroy() {
    this.clientPacketOns.clear();

    for (const cleanup of this.clientCleanups.values()) {
      cleanup();
    }

    this.clientCleanups.clear();
    this.clients.clear();

    this.cleanUpModules();

    this.scrubberSet.destroy();
  }

  /** Send the given message to all clients in this session */
  sendToAll(event: string, ...args: any[]) {
    for (const client of this.clients) {
      client.connection.send(event, ...args);
    }
  }

  /** Send the given message to all clients in this session except the given client */
  sendToOthers(self: NUsightSessionClient, event: string, ...args: any[]) {
    for (const client of this.clients) {
      if (client !== self) {
        client.connection.send(event, ...args);
      }
    }
  }

  /** Register a listener for a packet of the given event type from a client */
  onClientPacket(event: string, callback: ClientPacketCallback) {
    const callbacks = this.clientPacketOns.get(event) ?? new Set();

    callbacks.add(callback);

    this.clientPacketOns.set(event, callbacks);

    return () => {
      callbacks.delete(callback);
    };
  }

  /** Process a packet received from a client */
  processClientPacket(options: NUClearNetSend, client: NUsightSessionClient) {
    if (typeof options.type !== "string") {
      console.error("Unable to process client packet where type is not a string", options);
      return;
    }

    const callbacks = this.clientPacketOns.get(options.type);

    if (!callbacks) {
      return;
    }

    for (const callback of callbacks) {
      callback(client, options);
    }
  }
}
