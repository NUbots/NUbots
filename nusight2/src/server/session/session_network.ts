import { NUClearNetPacket } from "nuclearnet.js";
import { NUClearNetPeer } from "nuclearnet.js";

import { message, MessageType } from "../../shared/messages";
import { Emit } from "../../shared/messages/emit";
import { messageTypeToName } from "../../shared/messages/type_converters";
import { hashType } from "../nuclearnet/hash_type";

import { NUsightSession } from "./session";
import { NUsightSessionClient } from "./session_client";

import IRpcRequestMeta = message.eye.IRpcRequestMeta;
import IRpcResponseMeta = message.eye.IRpcResponseMeta;

export type NUClearMessageCallback<T> = (peer: NUClearNetPeer, message: T) => void;
export type ClientMessageCallback<T> = (client: NUsightSessionClient, message: T) => void;

export interface RpcResponse {
  rpc?: IRpcResponseMeta | null;
}
export type RpcRequestHandler<T> = (request: T, sender: NUsightSessionClient) => RpcResponse | Promise<RpcResponse>;

/** Used for messages from the current NUsight server */
const NUSIGHT_SERVER_PEER = { name: "nusight", address: "0.0.0.0", port: 0 } as const;

/**
 * Handles networking in a NUsight session: allows for exchanging messages between clients in the session and NUClearNet.
 */
export class NUsightSessionNetwork {
  /** The session this network belongs to */
  private session: NUsightSession;

  /** Create a new network for the given session */
  constructor(session: NUsightSession) {
    this.session = session;
  }

  /**
   * Emit the given Protobuf message. The target is chosen based on `options.target` as follows:
   *   - "nusight" or `undefined` -> send to all clients in the session
   *   - "nusight#<id>"           -> send to the client with the given id
   *   - "*"                      -> send to all clients and NUClearNet
   *   - "" or any other string   -> send to NUClearNet
   */
  emit: Emit = (message, options = {}) => {
    const messageType = message.constructor as MessageType<any>;
    const messageTypeName = messageTypeToName(messageType);

    const payload = messageType.encode(message).finish() as Buffer;
    const hash = hashType(messageTypeName);
    const reliable = options.reliable ?? true;
    const peer = NUSIGHT_SERVER_PEER;

    if (options.target === undefined || options.target === "nusight") {
      const packet: NUClearNetPacket = { hash, reliable, payload, peer };
      this.session.sendToAll(messageTypeName, packet);
    } else if (options.target.startsWith("nusight#")) {
      const id = Number(options.target.replace(/^nusight#/, ""));
      const client = Array.from(this.session.clients).find((client) => client.id === id);

      if (!client) {
        throw new Error(`Could not emit to NUsight client with id ${id}, no such client`);
      }

      const packet: NUClearNetPacket = { hash, reliable, payload, peer };
      client.connection.send(messageTypeName, packet);
    } else if (options.target === "*") {
      const packet: NUClearNetPacket = { hash, reliable, payload, peer };
      this.session.sendToAll(messageTypeName, packet);
      this.session.nuclearnetClient.send({ type: hash, payload, reliable });
    } else {
      this.session.nuclearnetClient.send({ type: hash, payload, reliable, target: options.target });
    }
  };

  /** Register a listener for the given message type from NUClearNet */
  onNUClearMessage<T>(
    type: MessageType<T> | { type: MessageType<T>; subtype?: number },
    cb: NUClearMessageCallback<T>,
  ) {
    const { event, MessageType } = findMessageType(type);

    return this.session.nuclearnetClient.on(event, (packet: NUClearNetPacket) => {
      const buffer = new Uint8Array(packet.payload);
      const message = MessageType.decode(buffer);
      cb(packet.peer, message);
    });
  }

  /** Register a listener for the given message type from a client in the session */
  onClientMessage<T>(type: MessageType<T> | { type: MessageType<T>; subtype?: number }, cb: ClientMessageCallback<T>) {
    const { event, MessageType } = findMessageType(type);

    return this.session.onClientPacket(event, (client, packet) => {
      const buffer = new Uint8Array(packet.payload);
      const message = MessageType.decode(buffer);
      cb(client, message);
    });
  }

  /** Register a handler to respond to RPC requests of the given type from a client in the session */
  onClientRpc<T extends { rpc?: IRpcRequestMeta | null }>(
    type: MessageType<T> | { type: MessageType<T>; subtype?: number },
    requestHandler: RpcRequestHandler<T>,
  ) {
    return this.onClientMessage(type, async (client: NUsightSessionClient, request: T) => {
      try {
        const response = await requestHandler(request, client);
        this.emit(response as any, { target: `nusight#${client.id}` });
      } catch (error: unknown) {
        const ResponseType = (request.constructor as any).Response;
        this.emit(
          new ResponseType({
            rpc: {
              ok: false,
              token: request.rpc?.token,
              error: error instanceof Error ? error.message : String(error),
            },
          }),
          { target: `nusight#${client.id}` },
        );
      }
    });
  }
}

/** Find the Protobuf message class and event name for the given type */
function findMessageType(type: MessageType<any> | { type: MessageType<any>; subtype?: number }) {
  const MessageType = typeof type === "object" ? type.type : type;
  const messageTypeName = messageTypeToName(MessageType);
  const subtype = typeof type === "object" ? type.subtype : null;

  return {
    event: `${messageTypeName}${subtype ? `#${subtype}` : ""}`,
    MessageType,
  };
}
