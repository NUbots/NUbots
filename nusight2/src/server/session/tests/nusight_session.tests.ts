import { NUClearNetPacket } from "nuclearnet.js";

import { message } from "../../../shared/messages";
import { makeScrubberStatePacket, sampleADefaultState } from "../../nbs_scrubber/tests/test_utils";
import { sampleFileA } from "../../nbs_scrubber/tests/test_utils";
import { scrubberPeerName } from "../../nbs_scrubber/tests/test_utils";
import { tick } from "../../nbs_scrubber/tests/test_utils";
import { hashType } from "../../nuclearnet/hash_type";
import { NUsightSession } from "../session";

import { createMockNUClearNetClient, createMockWebSocket, createPacketFromServer } from "./test_utils";

import Test = message.support.nusight.Test;
import ScrubberLoadRequest = message.eye.ScrubberLoadRequest;
import ScrubberCloseRequest = message.eye.ScrubberCloseRequest;
import ScrubberSeekRequest = message.eye.ScrubberSeekRequest;
import ScrubberClosed = message.eye.ScrubberClosed;
import { NUClearNetPeerWithType } from "../../../shared/nuclearnet/nuclearnet_client";

const testPacketType = "message.support.nusight.Test";

function createTestPacket(opts: { reliable?: boolean } = {}) {
  const payload = Test.encode({ message: "Test" }).finish();

  const packet: NUClearNetPacket = {
    hash: hashType(testPacketType),
    payload: Buffer.from(payload),
    reliable: opts.reliable ?? false,
    peer: {
      name: "nusight",
      address: "127.0.0.1",
      port: 0,
    },
  };

  return packet;
}

describe("NUsightSession and NUsightSessionClient", () => {
  let nuclearnetClient: ReturnType<typeof createMockNUClearNetClient>["nuclearnetClient"];
  let nuclearnetMockEmit: (event: string, ...args: any[]) => void;

  beforeEach(() => {
    ({ nuclearnetClient, nuclearnetMockEmit } = createMockNUClearNetClient());
  });

  it("forwards outgoing packets from a browser client to NUClearNet", () => {
    // Given we have a session with a mock client connected, ...
    const session = NUsightSession.of(nuclearnetClient);
    const mockSocket = createMockWebSocket();
    session.addClient(mockSocket.connection);

    // when the client sends a packet to the server, ...
    const packetFromClient = { type: "message.support.nusight.Test" };
    mockSocket.emit("packet", packetFromClient);

    // the server should forward the packet to NUClearNet.
    expect(nuclearnetClient.send).toHaveBeenCalledWith(packetFromClient);

    // Clean up
    session.destroy();
  });

  it("listens for incoming NUClearNet packets on behalf of a browser client", () => {
    // Given we have a session with a mock client connected, ...
    const session = NUsightSession.of(nuclearnetClient);
    const mockSocket = createMockWebSocket();
    session.addClient(mockSocket.connection);

    // and the client starts listening for a given packet type, ...
    mockSocket.emit("listen", testPacketType, "some-listen-token");

    // when a packet of that type is received from NUClearNet, ...
    const packet = createTestPacket();
    nuclearnetMockEmit(testPacketType, packet);

    // the packet should be forwarded to the client.
    expect(mockSocket.connection.send).toHaveBeenCalledWith(testPacketType, packet, expect.any(Function));

    // Ack the send to clear the timeout and end the test
    mockSocket.connection.send.mock.calls[0][2]();

    // Clean up
    session.destroy();
  });

  it("stops listening for incoming NUClearNet packets on request from browser client", () => {
    // Given we have a session with a mock client connected, ...
    const session = NUsightSession.of(nuclearnetClient);
    const mockSocket = createMockWebSocket();
    session.addClient(mockSocket.connection);

    // if the client starts listening for a given packet type ...
    mockSocket.emit("listen", testPacketType, "some-listen-token");

    // and then stops listening for that packet type, ...
    mockSocket.emit("unlisten", "some-listen-token");

    const packet = createTestPacket();
    nuclearnetMockEmit("message.support.nusight.Test", packet);

    // the packet should not be forwarded to the client since it's no longer listening
    expect(mockSocket.connection.send).not.toHaveBeenCalled();

    // Clean up
    session.destroy();
  });

  it("loads an nbs scrubber on valid scrubber load request from browser client", async () => {
    // Given we have a session with a mock client connected, ...
    const session = NUsightSession.of(nuclearnetClient);
    const mockSocket = createMockWebSocket();
    session.addClient(mockSocket.connection);

    // if the client starts listening for ScrubberState packets, ...
    mockSocket.emit("listen", "message.eye.ScrubberState", "some-listen-token");

    // and then requests to load a scrubber, ...
    const request = new ScrubberLoadRequest({ files: [sampleFileA], name: scrubberPeerName });
    mockSocket.emitMessage(request);

    await tick();

    const scrubberPeer: NUClearNetPeerWithType = {
      name: scrubberPeerName,
      address: "0.0.0.0",
      port: 1,
      type: "nbs-scrubber",
    };

    // a NUClear join message should be sent to the client to present the scrubber as a "peer", ...
    expect(mockSocket.connection.send).toHaveBeenNthCalledWith(1, "nuclear_join", scrubberPeer);

    const scrubberState = sampleADefaultState();

    // and a ScrubberState packet for the newly loaded scrubber should be sent.
    expect(mockSocket.connection.send).toHaveBeenNthCalledWith(
      2,
      "message.eye.ScrubberState",
      makeScrubberStatePacket({ ...scrubberState, timestamp: scrubberState.start, peer: scrubberPeer }),
      expect.any(Function),
    );

    // Clean up
    session.destroy();
  });

  it("sends loaded scrubbers to all clients in the session", async () => {
    // Given we have a session...
    const session = NUsightSession.of(nuclearnetClient);

    // and a client A connected to the session...
    const mockSocketA = createMockWebSocket();
    session.addClient(mockSocketA.connection);

    // and a client B connected to the session...
    const mockSocketB = createMockWebSocket();
    session.addClient(mockSocketB.connection);

    // and both clients start listening for ScrubberState packets...
    mockSocketA.emit("listen", "message.eye.ScrubberState", "listen-token-1");
    mockSocketB.emit("listen", "message.eye.ScrubberState", "listen-token-2");

    // when client A requests to load a scrubber...
    const request = new ScrubberLoadRequest({ rpc: { token: 1 }, files: [sampleFileA], name: scrubberPeerName });
    mockSocketA.emitMessage(request);

    await tick();

    const scrubberId = 1;
    const scrubberPeer: NUClearNetPeerWithType = {
      name: scrubberPeerName,
      address: "0.0.0.0",
      port: scrubberId,
      type: "nbs-scrubber",
    };
    const scrubberState = sampleADefaultState(scrubberPeerName);

    // a NUClear join event should be sent to both clients, ...
    expect(mockSocketA.connection.send).toHaveBeenNthCalledWith(1, "nuclear_join", scrubberPeer);
    expect(mockSocketB.connection.send).toHaveBeenNthCalledWith(1, "nuclear_join", scrubberPeer);

    // and a ScrubberState packet should be sent to both clients.
    expect(mockSocketA.connection.send).toHaveBeenNthCalledWith(
      2,
      "message.eye.ScrubberState",
      makeScrubberStatePacket({ ...scrubberState, timestamp: scrubberState.start, peer: scrubberPeer }),
      expect.any(Function),
    );
    expect(mockSocketB.connection.send).toHaveBeenNthCalledWith(
      2,
      "message.eye.ScrubberState",
      makeScrubberStatePacket({ ...scrubberState, timestamp: scrubberState.start, peer: scrubberPeer }),
      expect.any(Function),
    );

    // Clean up
    session.destroy();
  });

  it("sends all currently loaded scrubbers to a new browser client when it connects", async () => {
    // Given we have a session...
    const session = NUsightSession.of(nuclearnetClient);

    // and a client A connected to the session...
    const mockSocketA = createMockWebSocket();
    session.addClient(mockSocketA.connection);

    // when client A loads a scrubber...
    const request = new ScrubberLoadRequest({ rpc: { token: 1 }, files: [sampleFileA], name: scrubberPeerName });
    mockSocketA.emitMessage(request);

    const defaultState = sampleADefaultState(scrubberPeerName);
    const scrubberId = defaultState.id;
    const scrubberPeer: NUClearNetPeerWithType = {
      name: scrubberPeerName,
      address: "0.0.0.0",
      port: scrubberId,
      type: "nbs-scrubber",
    };

    // and seeks to a specific time...
    const seekRequest = new ScrubberSeekRequest({
      rpc: { token: 2 },
      id: scrubberId,
      timestamp: { seconds: 1149, nanos: 0 },
    });
    mockSocketA.emitMessage(seekRequest);

    await tick();

    // and client B joins...
    const mockSocketB = createMockWebSocket();
    session.addClient(mockSocketB.connection);

    // and starts listening for ScrubberState messages, ...
    mockSocketB.emit("listen", "message.eye.ScrubberState", "listen-token-2");

    // a NUClear join event should be sent to client B for the previously loaded scrubber, ...
    expect(mockSocketB.connection.send).toHaveBeenNthCalledWith(1, "nuclear_join", scrubberPeer);

    // along with a corresponding ScrubberState message.
    expect(mockSocketB.connection.send).toHaveBeenNthCalledWith(
      2,
      "message.eye.ScrubberState",
      makeScrubberStatePacket({ ...defaultState, timestamp: { seconds: 1149, nanos: 0 }, peer: scrubberPeer }),
      expect.any(Function),
    );

    // Clean up
    session.destroy();
  });

  it("responds with error on invalid scrubber load request from browser client", async () => {
    // Given we have a session with a mock client connected, ...
    const session = NUsightSession.of(nuclearnetClient);
    const mockSocket = createMockWebSocket();
    session.addClient(mockSocket.connection);

    // when the client requests to load a scrubber for an invalid file...
    const request = new ScrubberLoadRequest({ rpc: { token: 1 }, files: ["/file/that/does/not/exist.nbs"] });
    mockSocket.emitMessage(request);

    // the scrubber load should fail, with an error response sent back to the client.
    expect(mockSocket.connection.send).toHaveBeenCalledWith(
      "message.eye.ScrubberLoadRequest.Response",
      createPacketFromServer(
        new ScrubberLoadRequest.Response({
          rpc: {
            token: 1,
            ok: false,
            error: "Error: nbs index not found for file: /file/that/does/not/exist.nbs",
          },
        }),
      ),
    );

    // Clean up
    session.destroy();
  });

  it("handles scrubber seek requests from a browser client", async () => {
    // Given we have a session with a mock client connected, ...
    const session = NUsightSession.of(nuclearnetClient);
    const mockSocket = createMockWebSocket();
    session.addClient(mockSocket.connection);

    // when the client listen to a packet type, ...
    // (this tests that types registered before the scrubber is loaded are handled correctly)
    mockSocket.emit("listen", "message.Ping", "listen-token-ping");

    // and it loads a scrubber, ...
    const loadRequest = new ScrubberLoadRequest({ rpc: { token: 1 }, files: [sampleFileA] });
    mockSocket.emitMessage(loadRequest);

    // (ack the packet sent on scrubber load)
    mockSocket.connection.send.mock.calls[1][2](0, 0);
    await tick();

    // and then listens to another packet type, ...
    // (this tests that types registered after the scrubber is loaded are handled correctly)
    mockSocket.emit("listen", "message.Pong", "listen-token-pong");

    // (ack the packet sent from the Pong listen)
    mockSocket.connection.send.mock.calls[3][2](0, 0);
    await tick();

    // when it requests to seek to a specific time ...
    const updateRequest = new ScrubberSeekRequest({
      rpc: { token: 2 },
      id: 1,
      timestamp: { seconds: 1149, nanos: 0 },
    });
    mockSocket.emitMessage(updateRequest);

    await tick();

    // ... the scrubber should seek to the requested time and respond to the client with packets
    // at the seeked timestamp matching the packet types the client is currently listening for.
    // The `connection.send()` mock contains all the calls to send, with the arguments
    // of each call, so we match against a snapshot to make sure the right packets
    // were sent in the right order.
    expect(mockSocket.connection.send).toMatchSnapshot();

    // Clean up
    session.destroy();
  });

  it("closes a loaded scrubber on request from browser client", async () => {
    // Given we have a session with a mock clients A and B connected, ...
    const session = NUsightSession.of(nuclearnetClient);
    const mockSocketA = createMockWebSocket();
    const mockSocketB = createMockWebSocket();
    session.addClient(mockSocketA.connection);
    session.addClient(mockSocketB.connection);

    // and client A loads a scrubber, ...
    const loadRequest = new ScrubberLoadRequest({ rpc: { token: 1 }, files: [sampleFileA] });
    mockSocketA.emitMessage(loadRequest);
    await tick();

    const scrubberId = 1;

    // when client A closes the scrubber ...
    const stopRequest = new ScrubberCloseRequest({ rpc: { token: 2 }, id: scrubberId });
    mockSocketA.emitMessage(stopRequest);
    await tick();

    await mockSocketA.connection.send.waitForCall();

    // ... a NUClear leave event should be sent to both clients, ...
    const expectedLeavePeer: NUClearNetPeerWithType = {
      name: "sample-000-300.nbs",
      address: "0.0.0.0",
      port: scrubberId,
      type: "nbs-scrubber",
    };
    expect(mockSocketA.connection.send).toHaveBeenCalledWith("nuclear_leave", expectedLeavePeer);
    expect(mockSocketB.connection.send).toHaveBeenCalledWith("nuclear_leave", expectedLeavePeer);

    // ... a ScrubberClosed event should be sent to both clients, ...
    const expectedScrubberClosed = createPacketFromServer(new ScrubberClosed({ id: scrubberId }));
    expect(mockSocketA.connection.send).toHaveBeenCalledWith("message.eye.ScrubberClosed", expectedScrubberClosed);
    expect(mockSocketB.connection.send).toHaveBeenCalledWith("message.eye.ScrubberClosed", expectedScrubberClosed);

    // ... as well as a success response to the client that requested the scrubber to be closed.
    expect(mockSocketA.connection.send).toHaveBeenCalledWith(
      "message.eye.ScrubberCloseRequest.Response",
      createPacketFromServer(new ScrubberCloseRequest.Response({ rpc: { token: 2, ok: true } })),
    );

    // Attempting to seek the closed scrubber...
    const updateRequest = new ScrubberSeekRequest({
      rpc: { token: 3 },
      id: scrubberId,
      timestamp: { seconds: 1100, nanos: 0 },
    });
    mockSocketA.emitMessage(updateRequest);

    await tick();

    // ... should result in an error.
    expect(mockSocketA.connection.send).toHaveBeenCalledWith(
      "message.eye.ScrubberSeekRequest.Response",
      createPacketFromServer(
        new ScrubberSeekRequest.Response({
          rpc: {
            token: 3,
            ok: false,
            error: `scrubber ${scrubberId} not found for update (seek)`,
          },
        }),
      ),
    );

    // Clean up
    session.destroy();
  });
});
