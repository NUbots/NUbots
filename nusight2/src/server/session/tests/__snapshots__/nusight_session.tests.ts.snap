// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NUsightSession and NUsightSessionClient handles scrubber seek requests from a browser client 1`] = `
[MockFunction] {
  "calls": [
    [
      "nuclear_join",
      {
        "address": "0.0.0.0",
        "name": "sample-000-300.nbs",
        "port": 1,
        "type": "nbs-scrubber",
      },
    ],
    [
      "message.Ping",
      {
        "hash": {
          "data": [
            140,
            225,
            88,
            47,
            160,
            234,
            220,
            132,
          ],
          "type": "Buffer",
        },
        "payload": {
          "data": [
            112,
            105,
            110,
            103,
            46,
            48,
          ],
          "type": "Buffer",
        },
        "peer": {
          "address": "0.0.0.0",
          "name": "sample-000-300.nbs",
          "port": 1,
          "type": "nbs-scrubber",
        },
        "reliable": true,
      },
      [Function],
    ],
    [
      "message.eye.ScrubberLoadRequest.Response",
      {
        "hash": {
          "data": [
            238,
            58,
            171,
            90,
            248,
            96,
            89,
            3,
          ],
          "type": "Buffer",
        },
        "payload": {
          "data": [
            10,
            4,
            8,
            1,
            16,
            1,
          ],
          "type": "Buffer",
        },
        "peer": {
          "address": "0.0.0.0",
          "name": "nusight",
          "port": 0,
        },
        "reliable": true,
      },
    ],
    [
      "message.Pong",
      {
        "hash": {
          "data": [
            55,
            197,
            99,
            54,
            82,
            101,
            115,
            187,
          ],
          "type": "Buffer",
        },
        "payload": undefined,
        "peer": {
          "address": "0.0.0.0",
          "name": "sample-000-300.nbs",
          "port": 1,
          "type": "nbs-scrubber",
        },
        "reliable": true,
      },
      [Function],
    ],
    [
      "message.Ping",
      {
        "hash": {
          "data": [
            140,
            225,
            88,
            47,
            160,
            234,
            220,
            132,
          ],
          "type": "Buffer",
        },
        "payload": {
          "data": [
            112,
            105,
            110,
            103,
            46,
            52,
            57,
          ],
          "type": "Buffer",
        },
        "peer": {
          "address": "0.0.0.0",
          "name": "sample-000-300.nbs",
          "port": 1,
          "type": "nbs-scrubber",
        },
        "reliable": true,
      },
      [Function],
    ],
    [
      "message.Pong",
      {
        "hash": {
          "data": [
            55,
            197,
            99,
            54,
            82,
            101,
            115,
            187,
          ],
          "type": "Buffer",
        },
        "payload": {
          "data": [
            112,
            111,
            110,
            103,
            46,
            52,
            57,
          ],
          "type": "Buffer",
        },
        "peer": {
          "address": "0.0.0.0",
          "name": "sample-000-300.nbs",
          "port": 1,
          "type": "nbs-scrubber",
        },
        "reliable": true,
      },
      [Function],
    ],
    [
      "message.eye.ScrubberSeekRequest.Response",
      {
        "hash": {
          "data": [
            31,
            157,
            124,
            141,
            189,
            216,
            175,
            140,
          ],
          "type": "Buffer",
        },
        "payload": {
          "data": [
            10,
            4,
            8,
            2,
            16,
            1,
          ],
          "type": "Buffer",
        },
        "peer": {
          "address": "0.0.0.0",
          "name": "nusight",
          "port": 0,
        },
        "reliable": true,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
